src/backend/common/CommandBuffer.cpp

```diff
@@ -238,9 +257,75 @@ namespace backend {
             return true;
         };
 
+        // TODO(kainino@chromium.org): Manage this state inside an object, change lambda to a method
+        RenderPassBase* currentRenderPass = nullptr;
+        FramebufferBase* currentFramebuffer = nullptr;
+        uint32_t currentSubpass = 0;
+        auto beginSubpass = [&]() -> bool {
+            auto& subpassInfo = currentRenderPass->GetSubpassInfo(currentSubpass);
+            for (auto& kv : subpassInfo.colorAttachments) {
+                if (auto* tv = currentFramebuffer->GetTextureView(kv.second)) {
+                    auto* texture = tv->GetTexture();
+                    if (!texture->IsTransitionPossible(nxt::TextureUsageBit::ColorAttachment)) {
+                        device->HandleError("Can't transition attached texture to ColorAttachment usage");
+                        return false;
+                    }
+                    if (!textureHasGuaranteedUsageBit(texture, nxt::TextureUsageBit::ColorAttachment)) {
+                        mostRecentTextureUsages.erase(texture);
+                        texturesTransitioned.insert(texture);
+                    }
+                }
+            }
+            return true;
+        };
+
         Command type;
         while(iterator.NextCommandId(&type)) {
             switch (type) {
+                case Command::AdvanceSubpass:
+                    {
+                        iterator.NextCommand<AdvanceSubpassCmd>();
+                        if (currentRenderPass == nullptr) {
+                            device->HandleError("Can't advance subpass without an active render pass");
+                            return false;
+                        }
+                        if (currentSubpass + 1 >= currentRenderPass->GetSubpassCount()) {
+                            device->HandleError("Can't advance beyond the last subpass");
+                            return false;
+                        }
+                        currentSubpass += 1;
+                        if (!beginSubpass()) {
+                            return false;
+                        }
+                    }
+                    break;
+
+                case Command::BeginRenderPass:
+                    {
+                        BeginRenderPassCmd* begin = iterator.NextCommand<BeginRenderPassCmd>();
+                        auto* renderPass = begin->renderPass.Get();
+                        auto* framebuffer = begin->framebuffer.Get();
+                        if (currentRenderPass != nullptr) {
+                            device->HandleError("A render pass is already active");
+                            return false;
+                        }
+                        if (framebuffer->GetRenderPass() != renderPass) {
```

Comment:
Leave a TODO that this compatibility check will need to be slightly relaxed at some point, to allow render passes with different load and store ops.